text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1159 @@
+the performance impact of the schema conversions depend on the volume and type
+of transformations can be significant. We
+[benchmarked](https://github.com/tigrannajaryan/telemetry-schema/blob/main/schema/perf_test.go)
+one use-case where the telemetry schema conversion of spans and metrics from one
+version to another is performed using a Go implementation.
+
+The benchmark does the following:
+
+- Compares the CPU time necessary for schema conversion to the time necessary
+  for Protobuf decoding. This is a useful comparison because telemetry consumers
+  have to perform Protobuf decoding so it is the minimum baseline work against
+  which we can measure the impact of the additional conversion work.
+
+- Uses a hypothetical schema version change where 20 attributes in semantic
+  conventions were renamed.
+
+- Uses data composed of batches of 100 spans each with 10 attributes or 100
+  metric data points (Int64 Gauge type) with 2 labels per data point. Each batch
+  is associated with one resource that has 20 attributes.
+
+Here are the benchmark results:
+
+```
+BenchmarkDecode/Trace/Attribs-8             6121            919271 ns/op
+BenchmarkDecode/Metric/Int64-8              9516            635418 ns/op
+BenchmarkDecodeAndConvertSchema/Trace/Attribs-8  5988       943158 ns/op
+BenchmarkDecodeAndConvertSchema/Metric/Int64-8   8588       653266 ns/op
+```
+
+BenchmarkDecode is decoding time only. BenchmarkDecodeAndConvertSchema is
+decoding plus schema conversion time. We see that the processing time to do the
+conversion is about 3% of decoding time.
+
+The benchmarking done is for illustration purposes only. Real-world results will
+depend on the composition of the data, the volume of transformations done during
+conversion, programming language used, etc. However, we feel that this one data
+point is still useful for understanding that the potential impact is acceptable
+(and likely will be a very small percentage of total processing most telemetry
+consumers do).
+
+## Open Questions
+
+- Do we need to support a concept of "pre-release" semantic conventions that can
+  be broken freely and are not part of strict schema checks? Supposedly, this is
+  already possible simply by avoiding introducing a new schema version when such
+  semantic conventions change.
+
+- How to make it easy for instrumentation libraries to refer to a particular
+  OpenTelemetry schema version and also make sure any semantic convention
+  helpers the library uses (e.g. constants that define the semantic conventions)
+  match exactly that same schema version? One possible solution is to introduce
+  helper packages per schema version that the libraries can use, e.g. constants
+  that define the semantic conventions and the corresponding schema version url.
+  This should be likely the topic for a follow-up OTEP.
+
+- Should we make it possible to include the entire Schema File in OTLP requests
```

Comment:
[Thinking about how JSON Schema encourages re-use with a `$ref` keyword], I could imagine a "Yes" answer here that does not require any more protocol change than what you have here.

We could specify to use a schema URL like `resource://some.identifier` and then use the `some.identifier` resource value as a schema--that might work except for resource value size restrictions.

In a logging system, (maybe, if it were up to me, ...), I might suggest the process log its schema early on in the process lifetime and then use a schema URL that refers backwards in the log.

Reply:
I like this idea.

Reply:
Taking a note for the future. Resolving for now, since it is not mandatory for this initial proposal.